README.md
An autocompetion engine for Dart.

Supports highly configurable partial matching.

Fast and memory efficient.

Suitable for both web and mobile devices.

## Test It Live !

<mark>Play with a [Live instance of Suggester](https://derbec.github.io/suggester/).</mark>

Test different combinations of data and parameters tofind what suits your application.

## Usage

### Import

```dart
import 'package:suggester/suggester.dart';
```

### Choose Term Mapping

Suggester works by tokenising both suggestions and input using a Term Mapping and finding partial matches between them via prefix searching.

For example a TermMapping that groups inputs and suggestions into sequences of letters and numbers:

```dart
print(AlphaOrNumeric().map('Armani.Klein3@veda.io', false, false));
```

```shell
{armani, klein, 3, veda, io}
```

TermMapping that groups inputs and suggestions into [NGrams](https://en.wikipedia.org/wiki/N-gram) of size 3, i.e. [Trigrams](https://en.wikipedia.org/wiki/Trigram).


```dart
print(Ngrams(3).map('Armani.Klein3@veda.io', false, false));
```

```shell
{arm, rma, man, ani, ni., i.k, .kl, kle, lei, ein, in3, n3@, 3@v, @ve, ved, eda, da., a.i, .io}
```

### Construct Suggester and Load Entries

A small list of entries. Suggester comfortably handles any number.

```dart
final entries = [
  'Armani.Klein3@veda.io',
  'Tracy@garett.io',
  'Eula@shaniya.name',
  'Reynold.Kunde@leif.net',
  'Reanna@flavio.us',
  'Fidel@nelle.com',
  'Luciano_Pouros@keenan.ca',
  'Curtis_McGlynn@wendell.me',
  'Reese@dante.io',
  'Sallie3@ettie.me',
  'Yolanda@bertram.co.uk',
  'Mariam.Toy@oren.com',
  'Rosalia.Koepp@mellie.name',
  'Rowena.Klein@abdul.us',
  'Jeff@travon.me',
  'Mario_Zulauf@michael.com',
  'Amos@carolina.ca',
  'Lonny_Gleichner@stanton.co.uk',
  'Lula_Bartoletti@makenzie.org',
  'Hardy@marion.ca',
  'Libby_Parker@ursula.biz',
  'banjo.berwen@newell.ca',
  'Genesis@osvaldo.me',
  'Caesar@joana.biz',
  'Cathrine@shayne.info',
  'Bill@dejuan.org',
  'Victor_Crist@dina.ca',
  'Stephany@dillan.io',
  'Emmanuel.Padberg@newell.ca'
];
```

Construct 2 example Suggester objects, each using a different Term Mapping.
Load each with the above list of entries.

```dart
final suggesterAlphaOrNumeric = Suggester(AlphaOrNumeric())
  ..addAll(entries);

final suggesterTrigram = Suggester(Ngrams(3))..addAll(entries);

```

### Suggest!
Find all suggestions featuring search term 'wen', note the difference between TermMappings.

```dart
print(suggesterAlphaOrNumeric.suggest('wen').map((e) => e.entry.value));

```

```shell
(Curtis_McGlynn@wendell.me)

```

```dart
print(suggesterTrigram.suggest('wen').map((e) => e.entry.value));
```

```shell
(Rowena.Klein@abdul.us, banjo.berwen@newell.ca, Curtis_McGlynn@wendell.me)

```

### Mark Suggestion Search Terms

User feedback is helped by marking a suggestion with the search terms.
Pass 2 custom functions to be applied to marked and unmarked terms.
Terms can be mapped to any type, here it is mapped to html tags but could as well return an iterable of UI elements.

```dart
print(suggesterTrigram.suggest('wen').map((e) => e
    .mapTermsAndNonTerms((term) => '<mark>$term</mark>', (nonTerm) => nonTerm)
    .join()));

```

```shell
(Ro<mark>wen</mark>a.Klein@abdul.us, banjo.ber<mark>wen</mark>@newell.ca, Curtis_McGlynn@<mark>wen</mark>dell.me)

```

Ro<mark>wen</mark>a.Klein@abdul.us, banjo.ber<mark>wen</mark>@newell.ca, Curtis_McGlynn@<mark>wen</mark>dell.me

### Expand Search With Edit Distance

An edit distance may also be specified during search allowing terms to be matched using a [Hamming edit distance](https://en.wikipedia.org/wiki/Hamming_distance). This extends the range of a TermMapping. Using the original AlphaOrNumeric example with a maxEditDistance of 1:

```dart
print(suggesterAlphaOrNumeric
    .suggest('wen', maxEditDistance: 1)
    .map((e) => e.entry.value));

```

```shell
(Curtis_McGlynn@wendell.me, Genesis@osvaldo.me)
```

... and with a maxEditDistance of 2

```dart
for (final suggestion
    in suggesterAlphaOrNumeric.suggest('wen', maxEditDistance: 2)) {
  print(suggestion.entry.value);
}

```

```shell
Curtis_McGlynn@wendell.me
Genesis@osvaldo.me
Jeff@travon.me
Lonny_Gleichner@stanton.co.uk
Reanna@flavio.us
Reese@dante.io
Reynold.Kunde@leif.net
banjo.berwen@newell.ca
Bill@dejuan.org
Fidel@nelle.com
Yolanda@bertram.co.uk
Emmanuel.Padberg@newell.ca
Luciano_Pouros@keenan.ca
Rosalia.Koepp@mellie.name
Victor_Crist@dina.ca
Armani.Klein3@veda.io

```

### Accept Suggestions

Once a selection is accepted this can be recorded via the accept method:

```dart
final result = suggesterAlphaOrNumeric.suggest('lu');

print(result.map((final suggestion) => suggestion.entry.value));

```

Note ordering of search suggestions.

```shell
(Luciano_Pouros@keenan.ca, Lula_Bartoletti@makenzie.org)

```

Accept last suggestion in ordering:

```dart
result.last.entry.accept();

print(result.map((final suggestion) => suggestion.entry.value));

```

Last suggestion now becomes first to reflect prior acceptance.

```shell
(Lula_Bartoletti@makenzie.org, Luciano_Pouros@keenan.ca)

```

By default when a suggestion is accepted it results in a subScore value incrementing but any int value may specified including a timestamp so that last used suggestions are always favoured. For example:

```dart
result.last.entry.accept(subScore: DateTime.now().millisecondsSinceEpoch);

print(result.first.entry);

```

```shell
{value: Luciano_Pouros@keenan.ca, subScore: 1609579026000}

```

### Check Current Entries

Iterate through all current entries

```dart
print(suggesterAlphaOrNumeric.entries);
```

Iterate all entries ordered by subScore.
Note that the last accepted entry is first with timestamp subScore.

```dart
print(suggesterAlphaOrNumeric.entriesBySubScore());
```

```shell
[{value: Luciano_Pouros@keenan.ca, subScore: 1609569378812},
  {value: Lula_Bartoletti@makenzie.org, subScore: 1},
  {value: Amos@carolina.ca, subScore: 0},
  {value: Armani.Klein3@veda.io, subScore: 0},
  {value: Bill@dejuan.org, subScore: 0},
  {value: Caesar@joana.biz, subScore: 0},
  {value: Cathrine@shayne.info, subScore: 0},
  {value: Curtis_McGlynn@wendell.me, subScore: 0},
  {value: Emmanuel.Padberg@newell.ca, subScore: 0},
  {value: Eula@shaniya.name, subScore: 0},
  ...]
```

### Serialise
Suggester is serialisable so suggestions and approval history are easily stored and reused.


```dart

final json = suggesterAlphaOrNumeric.toJson();

final rehydrated = Suggester.fromJson(AlphaOrNumeric(), json);

if (SuggesterEquality().equals(rehydrated, suggesterAlphaOrNumeric)) {
  print('Rehydrated successfully!');
}

```

```shell
Rehydrated successfully!

```

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/derbec/suggester/issues


CHANGELOG.md
## 0.9.0-nullsafety.0

- Initial version

## 0.9.1-nullsafety.0

- Added createSuggestion method

## 0.9.2

- Updated dependencies. createSuggestion() now throws ArgumentError.

## 0.9.3

- Downgraded dependencies. Pub dev does not recognise latest.

## 0.9.4

- Improved suggestion weighting scheme to recognise matching token sequences between entries and search sentences.

## 0.9.5

- Fixed bug where initial search term ignored.

## 0.9.6

- Added setAllEntrySubScores()

## 0.9.7

- Upgraded ternarytreap version to latest


test/timing.dart
import 'dart:io';

import 'package:suggester/suggester.dart';
import 'package:suggester/term_mapping.dart';
import 'emails.dart';

void main(List<String> args) {
  final suggester = Suggester(Ngrams(3, padStart: true, padEnd: true));

  final timer = Stopwatch()..start();

  for (var i = 0; i < 5; i++) {
    suggester.addAll(emails);

    for (final email in emails) {
      suggester.add(String.fromCharCodes(email.codeUnits.reversed));
    }
  }
  timer.stop();
  stdout.writeln('${timer.elapsedMicroseconds}');
}


test/suggester_test.dart
import 'dart:convert';
import 'package:suggester/suggester.dart';
import 'package:suggester/term_mapping.dart';
import 'package:test/test.dart';
import 'emails.dart';

void main() {
  group('Suggester: ', () {
    setUp(() {});

    test('KeysMapping: Text', () {
      // Check term proximality calculated correctly
      final tokens = Tokens();

      expect([0, 1, 2, 9, 6, 7, 3].map((e) => tokens.tpFromTermIdx(e)), [
        1.0,
        0.9,
        0.81,
        0.3874204890000001,
        0.531441,
        0.4782969000000001,
        0.7290000000000001
      ]);

      expect(Tokens().map('This is a dog named, Banjo.', false, true),
          equals(['this', 'is', 'a', 'dog', 'named,', 'banjo.']));
      expect(Tokens().map('This is a dog named, Banjo.', true, true),
          equals(['This', 'is', 'a', 'dog', 'named,', 'Banjo.']));

      expect(Alpha().map('This is a dog named, Banjo.', false, true),
          equals(['this', 'is', 'a', 'dog', 'named', 'banjo']));
      expect(Alpha().map('banjo. thedog@gmail..thedog .com gmail', true, true),
          equals(['banjo', 'thedog', 'gmail', 'com']));
      expect(
          AlphaAndNumeric()
              .map('banjo. thedog@gmail..thedog test34 .com gmail', true, true),
          equals(['banjo', 'thedog', 'gmail', 'test34', 'com']));
      expect(
          AlphaOrNumeric()
              .map('ba4nj89o.thedog@g399m ail..thedog34.com*gmail', true, true),
          equals([
            'ba',
            '4',
            'nj',
            '89',
            'o',
            'thedog',
            'g',
            '399',
            'm',
            'ail',
            '34',
            'com',
            'gmail'
          ]));

      expect(
          Ngrams(2, padStart: true, padEnd: true)
              .map('This is a dog named, Banjo.', false, true),
          equals([
            '\u00A0t',
            'th',
            'hi',
            'is',
            's ',
            ' i',
            ' a',
            'a ',
            ' d',
            'do',
            'og',
            'g ',
            ' n',
            'na',
            'am',
            'me',
            'ed',
            'd,',
            ', ',
            ' b',
            'ba',
            'an',
            'nj',
            'jo',
            'o.',
            '.\u00A0'
          ]));

      expect(
          Ngrams(3, padStart: true, padEnd: true)
              .map('This is a dog named, Banjo.', false, true),
          equals([
            '\u00A0\u00A0t',
            '\u00A0th',
            'thi',
            'his',
            'is ',
            's i',
            ' is',
            's a',
            ' a ',
            'a d',
            ' do',
            'dog',
            'og ',
            'g n',
            ' na',
            'nam',
            'ame',
            'med',
            'ed,',
            'd, ',
            ', b',
            ' ba',
            'ban',
            'anj',
            'njo',
            'jo.',
            'o.\u00A0',
            '.\u00A0\u00A0'
          ]));

      expect(
          Ngrams(3, padStart: false, padEnd: true)
              .map('This is a dog named, Banjo.', false, true),
          equals([
            'thi',
            'his',
            'is ',
            's i',
            ' is',
            's a',
            ' a ',
            'a d',
            ' do',
            'dog',
            'og ',
            'g n',
            ' na',
            'nam',
            'ame',
            'med',
            'ed,',
            'd, ',
            ', b',
            ' ba',
            'ban',
            'anj',
            'njo',
            'jo.',
            'o.\u00A0',
            '.\u00A0\u00A0'
          ]));

      expect(
          Ngrams(3, padStart: true, padEnd: false)
              .map('This is a dog named, Banjo.', false, true),
          equals([
            '\u00A0\u00A0t',
            '\u00A0th',
            'thi',
            'his',
            'is ',
            's i',
            ' is',
            's a',
            ' a ',
            'a d',
            ' do',
            'dog',
            'og ',
            'g n',
            ' na',
            'nam',
            'ame',
            'med',
            'ed,',
            'd, ',
            ', b',
            ' ba',
            'ban',
            'anj',
            'njo',
            'jo.',
          ]));

      expect(Ngrams(3).map('dere', false, true), equals(['der', 'ere']));
      expect(Ngrams(2).map('dere', false, true), equals(['de', 'er', 're']));

      expect(
          Ngrams(2, padStart: false, padEnd: false)
              .map('This is a dog named, Banjo.', false, true),
          equals([
            'th',
            'hi',
            'is',
            's ',
            ' i',
            ' a',
            'a ',
            ' d',
            'do',
            'og',
            'g ',
            ' n',
            'na',
            'am',
            'me',
            'ed',
            'd,',
            ', ',
            ' b',
            'ba',
            'an',
            'nj',
            'jo',
            'o.'
          ]));

      expect(
          Ngrams(3).map('This is a dog named, Banjo.', false, true),
          equals([
            'thi',
            'his',
            'is ',
            's i',
            ' is',
            's a',
            ' a ',
            'a d',
            ' do',
            'dog',
            'og ',
            'g n',
            ' na',
            'nam',
            'ame',
            'med',
            'ed,',
            'd, ',
            ', b',
            ' ba',
            'ban',
            'anj',
            'njo',
            'jo.'
          ]));
    });

    test('addAll()', () async {
      final suggester = Suggester(Alpha());
      suggester.addAll(emails);
    });

    test('mapTermsAndNonTerms', () {
      var suggester = Suggester(Alpha());
      suggester.add('Derek_Barton@jarvis.com');

      var terms = suggester.mapTerms('o rt derek ere vis');

      var suggestion = suggester.suggestFromTerms(terms).first;

      var markedTerms = suggestion.mapTermsAndNonTerms<String>(
          (final term) => '<strong>' + term + '</strong>',
          (final term) => term);

      expect(
          markedTerms,
          equals([
            '<strong>Derek</strong>',
            '_Ba',
            '<strong>rto</strong>',
            'n@jar',
            '<strong>vis</strong>',
            '.com'
          ]));

      suggester = Suggester(Ngrams(3, padStart: true, padEnd: true));
      suggester.add('Derek_Barton@jarvis.com');

      terms = suggester.mapTerms('o rt derek ere vis');

      suggestion = suggester.suggestFromTerms(terms).first;

      markedTerms = suggestion.mapTermsAndNonTerms<String>(
          (final term) => '<strong>' + term + '</strong>',
          (final term) => term);

      expect(
          markedTerms,
          equals([
            '<strong>Derek</strong>',
            '_Ba',
            '<strong>rto</strong>',
            'n@jar',
            '<strong>vis</strong>',
            '.com'
          ]));
    });

    test('json', () {
      final suggester1 = Suggester(Alpha());
      suggester1.add('Banjo_Barton@hotmail.com');
      suggester1.add('Tilly_Smith@jarvis.com');
      suggester1.add('Derek_Barton@jarvis.com');

      final encoded = jsonEncode(suggester1.toJson());

      final suggester2 = Suggester.fromJson(Alpha(), jsonDecode(encoded));

      expect(SuggesterEquality().equals(suggester1, suggester2), equals(true));
    });

    test('SuggesterEquality', () {
      final suggester1 = Suggester(Alpha());
      suggester1.add('Banjo_Barton@hotmail.com');
      suggester1.add('Tilly_Smith@jarvis.com');
      suggester1.add('Derek_Barton@jarvis.com');
      suggester1.add('Tilly_Smith@jarvis.com');

      final suggester2 = Suggester(Alpha());
      suggester2.add('Tilly_Smith@jarvis.com');
      suggester2.add('Derek_Barton@jarvis.com');
      suggester2.add('Tilly_Smith@jarvis.com');
      suggester2.add('Banjo_Barton@hotmail.com');

      expect(SuggesterEquality().equals(suggester1, suggester2), equals(true));

      suggester1.suggest('Derek').first.entry.accept();

      expect(SuggesterEquality().equals(suggester1, suggester2), equals(false));
    });

    test('createSuggestion', () {
      final suggester1 = Suggester(Alpha());
      suggester1.add('Banjo_Barton@hotmail.com');
      suggester1.add('Tilly_Smith@jarvis.com');
      suggester1.add('Derek_Barton@jarvis.com');
      suggester1.add('Tilly_Smith@jarvis.com');

      final test = suggester1.createSuggestion('Tilly_Smith@jarvis.com');
      expect(test.entry.value, 'Tilly_Smith@jarvis.com');

      // Throw exception if unknown
      expect(() => suggester1.createSuggestion('Tilly_Smith@unkown.com'),
          throwsArgumentError);
    });
  });
}


lib/src/term_mappings.dart
import 'dart:collection';
import 'package:suggester/suggester.dart';

final RegExp _matchAlphaUnicode = RegExp(r'\p{L}+', unicode: true);
final RegExp _matchAlphaAscii = RegExp(r'[a-zA-Z]+');
final RegExp _matchAlphaAndNumericUnicode =
    RegExp(r'[\p{L}\p{Nl}\p{Nd}]+', unicode: true);
final RegExp _matchAlphaAndNumericAscii = RegExp(r'[0-9a-zA-Z]+');
final RegExp _matchAlphaOrNumericUnicode =
    RegExp(r'\p{L}+|[\p{Nl}\p{Nd}]+', unicode: true);
final RegExp _matchAlphaOrNumericAscii = RegExp(r'[0-9]+|[a-zA-Z]+');

final RegExp _matchSeparatorUnicode =
    RegExp(r'[\p{Zl}\p{Zp}\p{Zs}]+', unicode: true);
final RegExp _matchSeparatorAscii = RegExp(r'[\s]+');

String _collapseWhitespace(String str, bool unicode) => str
    .trim()
    .replaceAll(unicode ? _matchSeparatorUnicode : _matchSeparatorAscii, ' ');

/// Match runs of alphanumeric characters seperated by separators
class Tokens extends TermMapping {
  /// Construct tokens
  Tokens({double decay = 0.1})
      : super(
            (String str, bool caseSensitive, bool unicode) =>
                LinkedHashSet<String>.from((caseSensitive
                        ? str
                        : str.toLowerCase())
                    .split(
                        unicode ? _matchSeparatorUnicode : _matchSeparatorAscii)
                    .where((final term) => term.isNotEmpty)),
            decay);
}

/// Each term represents a run of Letters
class Alpha extends TermMapping {
  /// Construct tokens
  Alpha({double decay = 0.1})
      : super(
            (String str, bool caseSensitive, bool unicode) =>
                LinkedHashSet<String>.from(
                    (unicode ? _matchAlphaUnicode : _matchAlphaAscii)
                        .allMatches(caseSensitive ? str : str.toLowerCase())
                        .map((m) => m[0])),
            decay);
}

/// Each term represents a run of Letters and Numbers
class AlphaAndNumeric extends TermMapping {
  /// Construct tokens
  AlphaAndNumeric({double decay = 0.1})
      : super(
            (String str, bool caseSensitive, bool unicode) =>
                LinkedHashSet<String>.from((unicode
                        ? _matchAlphaAndNumericUnicode
                        : _matchAlphaAndNumericAscii)
                    .allMatches(caseSensitive ? str : str.toLowerCase())
                    .map((m) => m[0])),
            decay);
}

/// Each term represents a run of Letters or Numbers
class AlphaOrNumeric extends TermMapping {
  /// Construct tokens
  AlphaOrNumeric({double decay = 0.1})
      : super(
            (String str, bool caseSensitive, bool unicode) =>
                LinkedHashSet<String>.from((unicode
                        ? _matchAlphaOrNumericUnicode
                        : _matchAlphaOrNumericAscii)
                    .allMatches(caseSensitive ? str : str.toLowerCase())
                    .map((m) => m[0])),
            decay);
}

/// Create ngrams of specified size and padding character
class Ngrams extends TermMapping {
  /// Construct [Ngrams] with gram size.
  Ngrams(int n,
      {double decay = 0.1,
      bool padStart = false,
      bool padEnd = false,
      String padChar = '\u00A0'})
      : super((String str, bool caseSensitive, bool unicode) {
          padStart = padStart;
          padEnd = padEnd;
          padChar = padChar;
          final orderedNgrams =
              LinkedHashSet<String>(); // ignore: prefer_collection_literals

          if (!caseSensitive) {
            str = str.toLowerCase();
          }

          str = _collapseWhitespace(str, unicode);

          final strLen = str.length;

          final ngramBuffer = StringBuffer();

          final start = padStart ? -(n - 1) : 0;
          final end = padEnd ? strLen : strLen - (n - 1);

          for (var i = start; i < end; i++) {
            for (var j = 0; j < n; j++) {
              final strIdx = i + j;
              if (strIdx < 0 || strIdx >= strLen) {
                ngramBuffer.write(padChar);
              } else {
                ngramBuffer.write(str[strIdx]);
              }
            }
            orderedNgrams.add(ngramBuffer.toString());
            ngramBuffer.clear();
          }

          return orderedNgrams;
        }, decay);
}


example/suggester_example.dart
// ignore_for_file: avoid_print

import 'package:suggester/suggester.dart';
import 'package:suggester/term_mapping.dart';

void main() {
  // A small list of entries. Suggester comfortably handles any number.
  final entries = [
    'Armani.Klein3@veda.io',
    'Tracy@garett.io',
    'Eula@shaniya.name',
    'Reynold.Kunde@leif.net',
    'Reanna@flavio.us',
    'Fidel@nelle.com',
    'Luciano_Pouros@keenan.ca',
    'Curtis_McGlynn@wendell.me',
    'Reese@dante.io',
    'Sallie3@ettie.me',
    'Yolanda@bertram.co.uk',
    'Mariam.Toy@oren.com',
    'Rosalia.Koepp@mellie.name',
    'Rowena.Klein@abdul.us',
    'Jeff@travon.me',
    'Mario_Zulauf@michael.com',
    'Amos@carolina.ca',
    'Lonny_Gleichner@stanton.co.uk',
    'Lula_Bartoletti@makenzie.org',
    'Hardy@marion.ca',
    'Libby_Parker@ursula.biz',
    'banjo.berwen@newell.ca',
    'Genesis@osvaldo.me',
    'Caesar@joana.biz',
    'Cathrine@shayne.info',
    'Bill@dejuan.org',
    'Victor_Crist@dina.ca',
    'Stephany@dillan.io',
    'Emmanuel.Padberg@newell.ca'
  ];

  // Suggester works by tokenising both suggestions and input using a Term Mapping
  // and finding partial matches between them via prefix searching.
  // For example a TermMapping that groups inputs and suggestions into sequences of
  // letters and numbers:
  print(AlphaOrNumeric().map('Armani.Klein3@veda.io', false, false));

  // {armani, klein, 3, veda, io}

  // TermMapping that groups inputs and suggestions into [NGrams](https://en.wikipedia.org/wiki/N-gram)
  // of size 3, i.e. trigrams.
  print(Ngrams(3).map('Armani.Klein3@veda.io', false, false));

  // {arm, rma, man, ani, ni., i.k, .kl, kle, lei, ein, in3, n3@, 3@v, @ve, ved, eda, da., a.i, .io}

  // Once you decide on a TermMapping construct a Suggester and add suggestions:
  final suggesterAlphaOrNumeric = Suggester(AlphaOrNumeric())..addAll(entries);
  final suggesterTrigram = Suggester(Ngrams(3))..addAll(entries);

  // Find all suggestions featuring search term 'wen', note the difference between tokenisers.
  print(suggesterAlphaOrNumeric.suggest('wen').map((e) => e.entry.value));

  // (Curtis_McGlynn@wendell.me)

  print(suggesterTrigram.suggest('wen').map((e) => e.entry.value));

  // (Rowena.Klein@abdul.us, banjo.berwen@newell.ca, Curtis_McGlynn@wendell.me)

  // You may want to mark the results with the search terms for user feedback:
  // Pass 2 custom functions to be applied to marked and unmarked.
  // Terms can be mapped to any type, here it is mapped to <em> tags but could as well
  // return an iterable of UI elements.
  print(suggesterTrigram.suggest('wen').map((e) => e
      .mapTermsAndNonTerms((term) => '<em>$term</em>', (nonTerm) => nonTerm)
      .join()));

  // ([Ro, <em>wen</em>, a.Klein@abdul.us], [banjo.ber, <em>wen</em>, @newell.ca], [Curtis_McGlynn@, <em>wen</em>, dell.me])

  // An edit distance may also be specified during search allowing terms to be
  // matched using a Hamming edit distance. This extends the range of
  // TermMapping. Using the original AlphaOrNumeric example with a
  // maxEditDistance of 1:
  print(suggesterAlphaOrNumeric
      .suggest('wen', maxEditDistance: 1)
      .map((e) => e.entry.value));

  // (Curtis_McGlynn@wendell.me, Genesis@osvaldo.me)

  // ... and with a maxEditDistance of 2
  for (final suggestion
      in suggesterAlphaOrNumeric.suggest('wen', maxEditDistance: 2)) {
    print(suggestion.entry.value);
  }

  // Curtis_McGlynn@wendell.me
  // Genesis@osvaldo.me
  // Jeff@travon.me
  // Lonny_Gleichner@stanton.co.uk
  // Reanna@flavio.us
  // Reese@dante.io
  // Reynold.Kunde@leif.net
  // banjo.berwen@newell.ca
  // Bill@dejuan.org
  // Fidel@nelle.com
  // Yolanda@bertram.co.uk
  // Emmanuel.Padberg@newell.ca
  // Luciano_Pouros@keenan.ca
  // Rosalia.Koepp@mellie.name
  // Victor_Crist@dina.ca
  // Armani.Klein3@veda.io

  // Once a selection is accepted this can be recorded via the accept method:
  final result = suggesterAlphaOrNumeric.suggest('lu');

  // Note ordering of search suggestions
  print(result.map((final suggestion) => suggestion.entry.value));
  //(Luciano_Pouros@keenan.ca, Lula_Bartoletti@makenzie.org)

  // Accept last suggestion in ordering
  result.last.entry.accept();

  // Last suggestion now becomes first to reflect prior acceptance.
  print(result.map((final suggestion) => suggestion.entry.value));

  // (Lula_Bartoletti@makenzie.org, Luciano_Pouros@keenan.ca)

  // By default when a suggestion is accepted it results in
  // a subScore value incrementing but any int value may speified
  // including a timestamp so that last used suggestions are always favoured.
  // for example:

  result.last.entry.accept(subScore: DateTime.now().millisecondsSinceEpoch);

  print(result.first.entry);

  // {value: Luciano_Pouros@keenan.ca, subScore: 1609579026000}

  // Iterate through all current entries
  print(suggesterAlphaOrNumeric.entries);

  // Iterate all entries ordered by subScore.
  // Note that the last accepted entry is first with timestamp subScore
  print(suggesterAlphaOrNumeric.entriesBySubScore());

  // [{value: Luciano_Pouros@keenan.ca, subScore: 1609569378812},
  // {value: Lula_Bartoletti@makenzie.org, subScore: 1},
  // {value: Amos@carolina.ca, subScore: 0},
  // {value: Armani.Klein3@veda.io, subScore: 0},
  // {value: Bill@dejuan.org, subScore: 0},
  // {value: Caesar@joana.biz, subScore: 0},
  // {value: Cathrine@shayne.info, subScore: 0},
  // {value: Curtis_McGlynn@wendell.me, subScore: 0},
  // {value: Emmanuel.Padberg@newell.ca, subScore: 0},
  // {value: Eula@shaniya.name, subScore: 0},
  // ...]

  // Suggester is serialisable so suggestions and approval history are easily store
  final json = suggesterAlphaOrNumeric.toJson();

  final rehydrated = Suggester.fromJson(AlphaOrNumeric(), json);

  // ... and rehydrated later
  if (SuggesterEquality().equals(rehydrated, suggesterAlphaOrNumeric)) {
    print('Rehydrated successfully!');
  }

  // Rehydrated successfully!
}
